upcloud/service/retry.go

```diff
@@ -0,0 +1,44 @@
+package service
+
+import (
+	"context"
+	"time"
+)
+
+type retryConfig struct {
+	interval time.Duration
+	// Inverse the should retry logic. By default, operation is retried until operation returns a value. If inverse is set to true, operation is retried while operation returns a value. This should be used, for example, for waiting until resource is deleted.
+	inverse bool
+}
+
+var defaultConfig = retryConfig{
+	interval: time.Second * 5,
+	inverse:  false,
+}
+
+func retry[T any](ctx context.Context, operation func(int, context.Context) (*T, error), config *retryConfig) (*T, error) {
+	if config == nil {
+		config = &defaultConfig
+	}
+
+	ticker := time.NewTicker(config.interval)
```

Comment:
Validate that interval is not negative, see [TF acceptance test error](https://github.com/UpCloudLtd/terraform-provider-upcloud/actions/runs/7656289645/job/20864777953#step:5:313)